zookeeper-server/src/main/java/org/apache/zookeeper/ClientCnxn.java

```diff
@@ -1239,6 +1238,7 @@ public void run() {
                             clientCnxnSocket.getIdleRecv(),
                             Long.toHexString(sessionId));
                         LOG.warn(warnInfo);
+                        changeZkState(States.CLOSED);
```

Comment:
Seems that the `to` could be simple per connection timeout. `readTimeout` and `connectTimeout` are apparently different.